src/config/Config.h

```diff
@@ -0,0 +1,195 @@
+#ifndef CASBIN_CPP_CONFIG_CONFIG
+#define CASBIN_CPP_CONFIG_CONFIG
+
+#include <string>
+#include <sstream>
+#include <unordered_map>
+#include <mutex>
+#include <fstream>
+#include <stdio.h>
+#include <algorithm>
+
+#include "./ConfigInterface.h"
+#include "../exception/IOException.h"
+#include "../exception/IllegalArgumentException.h"
+#include "../util/trim.h"
+#include "../util/ends_with.h"
+#include "../util/split.h"
+
+#endif
+
+using namespace std;
+
+mutex mtx_lock;
+
+class Config : public ConfigInterface {
+    private: 
+        static const string DEFAULT_SECTION;
+        static const string DEFAULT_COMMENT;
+        static const string DEFAULT_COMMENT_SEM;
+        unordered_map < string, unordered_map <string, string> > data;
```

Comment:
1. You should have used a **shared_mutex** instead of simple **mutex**, because the former is more similar and efficient to the **RWMutex** used in **Go**.
2. The mutex needs to be a member of the **ConfigInterface** instead of being defined globally, as done in Go implementation. Because if a single global mutex is used for all different instances of **Config**, then concurrency will decrease.

Reply:
Yeah, you are right, shared_mutex is more efficient in this case and of more similar to the go's core casbin implementation, but shared_mutex was included in c++17 hence, due to the tradeoff had to go with mutex instead which seems to perform good enough. Cannot compromise with the compiler support.

Reply:
Implementation has to be supported on C++11.

Reply:
You have a point, still you can consider to have mutex as the member of class if you would like to. Also, @hsluoyz if you please tell if features of C++17 or maybe C++14 can be used.